internal/internal.go

```diff
@@ -2,201 +2,394 @@ package internal
-// add the duration to a summed timer in the parent event.
-func rollup(ctx context.Context, ev *libhoney.Event, dur float64) {
-	parentEv := beeline.ContextEvent(ctx)
-	if parentEv == nil {
-		return
+// StartSpanWithEvent lets you take an event you've created outside the beeline
+// and push it in to the trace. This function will assign a parent, span, and
+// trace ID to the event and slot it in to the trace.
+func StartSpanWithEvent(ctx context.Context, ev *libhoney.Event) (context.Context, *Span) {
+	var span *Span
+	ctx, span = StartSpan(ctx, "")
+	span.ev = ev
+	return ctx, span
+}
+
+func StartAsyncSpan(ctx context.Context, name string) (context.Context, *Span) {
+	var span *Span
+	ctx, span = StartSpan(ctx, "")
+	span.amAsync = true
+	return ctx, span
+}
+
+func StartTraceWithIDs(ctx context.Context, traceID, parentID, name string) (context.Context, *Span) {
+	span := MakeNewTrace(traceID, parentID, name)
+	ctx = PutCurrentSpanInContext(ctx, span)
+	ctx = PutTraceInContext(ctx, span.trace)
+	return ctx, span
+}
+
+func MakeNewTrace(traceID, parentID, name string) *Span {
+	if traceID == "" {
+		tid, _ := uuid.NewRandom()
+		traceID = tid.String()
 	}
-	// ok now parentEv exists. lets add this to a total duration for the
-	// meta.type and the specific db call
-	evFields := ev.Fields()
-	pvFields := parentEv.Fields()
-
-	// only roll up if we have a meta type
-	metaType, ok := evFields["meta.type"]
-	if ok {
-		// make our field names
-		totalMetaCountKey := fmt.Sprintf("totals.%s.count", metaType)
-		totalMetaDurKey := fmt.Sprintf("totals.%s.duration_ms", metaType)
-		// get the existing values or zero if they're missing
-		totalTypeCount, _ := pvFields[totalMetaCountKey]
-		totalTypeCountVal, ok := totalTypeCount.(int)
-		if !ok {
-			totalTypeCountVal = 0
-		}
-		totalTypeDur, _ := pvFields[totalMetaDurKey]
-		totalTypeDurVal, ok := totalTypeDur.(float64)
-		if !ok {
-			totalTypeDurVal = 0
+	sid, _ := uuid.NewRandom()
+	spanID := sid.String()
+
+	trace := &Trace{
+		headers: TraceHeader{
+			TraceID: traceID,
+		},
+		spans:            make([]*Span, 0, 2), // most traces will have at least 2 spans
+		traceLevelFields: make(map[string]interface{}),
+		rollupFields:     make(map[string]float64),
+	}
+	// if a deterministic sampler is defined, use it. Otherwise sampling happens
+	// via the hook at send time.
+	var shouldDrop bool
+	var sampleRate = 1
+	if sample.GlobalSampler != nil {
+		shouldDrop = !sample.GlobalSampler.Sample(traceID)
+		sampleRate = sample.GlobalSampler.GetSampleRate()
+	}
+	trace.shouldDrop = shouldDrop
+	trace.sampleRate = sampleRate
+
+	span := &Span{
+		shouldDrop: shouldDrop,
+		timer:      timer.Start(),
+		amRoot:     true,
+		trace:      trace,
+		traceID:    traceID,
+		spanID:     spanID,
+		parentID:   parentID,
+		ev:         libhoney.NewEvent(),
+	}
+	span.ev.SampleRate = uint(trace.sampleRate)
+	span.ev.AddField("name", name)
+	span.ev.AddField("meta.root_span", true)
```

Comment:
Say that service A starts a trace, and makes a call to service B, with trace headers. Service B calls `MakeNewTrace` with the IDs from the trace headers. Then the new span in service B has `meta.root_span: true`, even though it's not the global root span of the trace. It's only the root of the subtrace in service B. Do we care?